src/mongocxx/test/database.cpp

```diff
@@ -492,4 +492,60 @@ TEST_CASE("Database integration tests", "[database]") {
     }
 }
 
+// As C++11 lacks generic lambdas, and "ordinary" templates can't appear at block scope,
+// we'll have to define our helper for the serviceId tests here. This implementation would
+// be more straightforward in newer versions of C++ (C++14 and on allow generic lambdas), 
+// see CXX-2350 (migration to more recent C++ standards):
+template <typename EventT>
+struct check_service_id
+{
+    void operator()(const EventT& event) {
+
+    // We MUST NOT report a service_id() outside of load-balancing mode, but in
+    // load-balancing mode, a value is required:
+    INFO("checking for service_id()")
+    CAPTURE(event.command_name(), test_util::is_load_balanced());
+    auto service_id = event.service_id();
+    if (test_util::is_load_balanced())
```

Comment:
The environment variable `MONGOC_TEST_LOADBALANCED` is currently not set. `service_id` will only be present exactly when the corresponding  C driver function is mocked to include it.

I suggest adding a field to this struct, like `bool expect_service_id` which the test can set based on whether the corresponding C driver function is mocked.

Reply:
I used a feature of Catch2 to generate positive and negative test cases in combination with a helper. A bit odd, but I think somewhat better than just duplicating all the code to get a "not".